back/engines/commercial/multi_tenancy/lib/tasks/core/insert_option_key_value_in_user_custom_field_values.rake

```diff
@@ -0,0 +1,67 @@
+# This rake task is designed specifically to insert custom_field key: custom_field_option key value pairs
+# into the user.custom_field_values, if given a value for the user matches (after sanitization) an
+# existing related custom_field_option multiloc value.
+# The task is designed only for single-locale tenants.
+#
+# The planned usage is for re-inserting values after a custom_field of a text registration question has been removed
+# and replaced with a single-select option question. Typically, this would be used for postcode-like values.
+#
+# The task assumes that the values have been exported from the database, probably before the related custom_field was
+# deleted, which would result in the loss of the respective data from the user.custom_field_values.
+#
+# The task always uses the first locale in the core locales list as the locale for the custom_field_option titles.
+# This will probably work even on multi-locale tenants when the options are postcode-like values, but care should
+# be taken when using this task on multi-locale tenants, and local testing is recommended.
+#
+# A dry run can be performed by passing anything other than 'execute' as the last argument, or by omitting it.
+#
+# Example:
+# Given a CSV file with the following headers: 'id,value', where id is the User ID and value is the value
+# to be checked against existing custom_field_option values.
+#
+# $ rake cl2_back:insert_option_key_values_in_user_custom_field_values['/user_values.csv','cqc.citizenlab.co','custom_field_id','execute']
+namespace :cl2_back do
+  desc 'Insert option key-value pairs to user custom_field_values hashes.'
+  task :insert_option_key_values_in_user_custom_field_values, %i[url host custom_field_id execute] => [:environment] do |_t, args|
+    data = CSV.parse(open(args[:url]).read, headers: true, col_sep: ',', converters: [])
+    custom_field = CustomField.find_by(id: args[:custom_field_id])
+
+    if custom_field.nil?
+      puts "ERROR: Couldn't find custom_field with id #{args[:custom_field_id]}"
+      next
+
+    end
+
+    options = custom_field.options
+    execute = args[:execute] == 'execute'
+    locale = AppConfiguration.instance.settings.dig('core', 'locales').first
+    count = 0
+
+    puts 'DRY RUN' unless execute
+
+    Apartment::Tenant.switch(args[:host].tr('.', '_')) do
+      data.each do |d|
+        next if d['value'].nil?
+
+        user = User.find_by(id: d['id'])
+        next if user.nil? || user.custom_field_values[custom_field.key].present?
+
+        value = d['value'].delete(' ').downcase
```

Comment:
Nit: I think `strip` could be more useful than `delete(' ')`?

Reply:
AFAIK `strip` removes leading and trailing whitespace, which indeed might clean up a few more than just `delete`.

I'm using `delete`, however, because often postcodes are typed with spaces, eg `'1234 AX'`, whereas the options have no spaces. I could use `gsub` or `tr` for this also.

Reply:
Ah right, I thought you wouldn't want to remove spaces in the middle, but so I was wrong then :)